// folio/out/index.js
"use strict";
/**
 * Copyright 2019 Google Inc. All rights reserved.
 * Modifications copyright (c) Microsoft Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.reporters = exports.merge = exports.newTestType = exports.globalTeardown = exports.globalSetup = exports.setReporters = exports.setConfig = exports.expect = void 0;
const spec_1 = require("./spec");
const dot_1 = __importDefault(require("./reporters/dot"));
const json_1 = __importDefault(require("./reporters/json"));
const junit_1 = __importDefault(require("./reporters/junit"));
const line_1 = __importDefault(require("./reporters/line"));
const list_1 = __importDefault(require("./reporters/list"));
__exportStar(require("./types"), exports);
var expect_1 = require("./expect");
Object.defineProperty(exports, "expect", { enumerable: true, get: function () { return expect_1.expect; } });
var spec_2 = require("./spec");
Object.defineProperty(exports, "setConfig", { enumerable: true, get: function () { return spec_2.setConfig; } });
Object.defineProperty(exports, "setReporters", { enumerable: true, get: function () { return spec_2.setReporters; } });
Object.defineProperty(exports, "globalSetup", { enumerable: true, get: function () { return spec_2.globalSetup; } });
Object.defineProperty(exports, "globalTeardown", { enumerable: true, get: function () { return spec_2.globalTeardown; } });
function newTestType() {
    return spec_1.newTestTypeImpl();
}
exports.newTestType = newTestType;
function merge(...envs) {
    return spec_1.mergeEnvsImpl(envs);
}
exports.merge = merge;
exports.reporters = {
    dot: dot_1.default,
    json: json_1.default,
    junit: junit_1.default,
    line: line_1.default,
    list: list_1.default,
};
